apps/heft/src/plugins/CopyFilesPlugin.ts

```diff
@@ -0,0 +1,265 @@
+    logger.terminal.writeLine(
+      `Copied ${copyCount} file${copyCount === 1 ? '' : 's'} and linked ${hardlinkCount} ` +
+        `file${hardlinkCount === 1 ? '' : 's'} in ${Math.round(duration)}ms`
+    );
+  }
+
+  protected async copyFilesAsync(copyDescriptors: ICopyFileDescriptor[]): Promise<[number, number]> {
+    if (copyDescriptors.length === 0) {
+      return [0, 0];
+    }
+
+    let copyCount: number = 0;
+    let hardlinkCount: number = 0;
+    await Async.forEachLimitAsync(
+      copyDescriptors,
+      MAX_PARALLELISM,
+      async (copyDescriptor: ICopyFileDescriptor) => {
+        if (copyDescriptor.hardlink) {
+          // Hardlink doesn't allow passing in overwrite param, so delete ourselves
+          try {
+            await FileSystem.deleteFileAsync(copyDescriptor.destinationFilePath, { throwIfNotExists: true });
+          } catch (e) {
+            if (!FileSystem.isFileDoesNotExistError(e)) {
+              throw e;
+            }
+            // Since the file doesn't exist, the parent folder may also not exist
+            await FileSystem.ensureFolderAsync(path.dirname(copyDescriptor.destinationFilePath));
+          }
+
+          await FileSystem.createHardLinkAsync({
+            linkTargetPath: copyDescriptor.sourceFilePath,
+            newLinkPath: copyDescriptor.destinationFilePath
+          });
+          hardlinkCount++;
+        } else {
+          // If it's a copy, simply call the copy function
+          await FileSystem.copyFileAsync({
+            sourcePath: copyDescriptor.sourceFilePath,
+            destinationPath: copyDescriptor.destinationFilePath,
+            alreadyExistsBehavior: AlreadyExistsBehavior.Overwrite
+          });
+          copyCount++;
+        }
+      }
+    );
+
+    return [copyCount, hardlinkCount];
+  }
+
+  private async _getCopyFileDescriptorsAsync(
+    buildFolder: string,
+    copyConfigurations: IExtendedSharedCopyConfiguration[]
+  ): Promise<ICopyFileDescriptor[]> {
+    // Create a map to deduplicate and prevent double-writes
+    // resolvedDestinationFilePath -> [resolvedSourceFilePath, hardlink]
+    const destinationCopyDescriptors: Map<string, ICopyFileDescriptor> = new Map();
+
+    for (const copyConfiguration of copyConfigurations) {
+      // Resolve the source folder path which is where the glob will be run from
+      const resolvedSourceFolderPath: string = path.resolve(buildFolder, copyConfiguration.sourceFolder);
+
+      // Glob extensions with a specific glob to increase perf
+      let sourceFileRelativePaths: Set<string>;
+      if (copyConfiguration.fileExtensions?.length) {
+        const escapedExtensions: string[] = globEscape(copyConfiguration.fileExtensions);
+        const pattern: string = `**/*+(${escapedExtensions.join('|')})`;
+        sourceFileRelativePaths = await this._expandGlobPatternAsync(
+          resolvedSourceFolderPath,
+          pattern,
+          copyConfiguration.excludeGlobs
+        );
```

Comment:
I wonder if this would be faster if we crawled the filesystem explicitly and looked for files matching one of the extensions instead of constructing and evaluating a glob.

Reply:
Switched to fast-glob. We talked offline, but under the covers this uses `micromatch` which is what I was planning on using anyway, alongside a manual crawl of the directory. We're also able to do a single call with multiple different glob patterns in one go, avoiding enumerating the filesystem multiple times when running patterns in the same folder. Ideally this should have the same perf as a custom solution, save for the additional calls we would have to make when targeting other `sourceFolder`s

Another thought here would be that we could combine all globs across all copy configurations, while appending the individual `sourceFolder` path to the glob, though I'm unsure if this would cause a perf hit, ie:
```json
{
  "sourceFolder": "src/subfolder",
  "includeGlobs": [ "**/*.json", "**/*.md" ] // -> "includeGlobs": [ "src/subfolder/**/*.json", "src/subfolder/**/*.md" ]
  ...
}
```